lib/grizzly/trace.ex

```diff
@@ -133,41 +129,60 @@ defmodule Grizzly.Trace do
   def init(opts) do
     size = Keyword.get(opts, :size, @default_size)
     record_keepalives = Keyword.get(opts, :record_keepalives, true)
-    {:ok, %{buffer: CircularBuffer.new(size), record_keepalives: record_keepalives}}
+    {:ok, %{buffers: %{}, size: size, record_keepalives: record_keepalives}}
   end
 
   @impl GenServer
-  def handle_cast({:log, <<0x23, 0x03, _ack_flag>>, _opts}, %{record_keepalives: false} = state) do
+  def handle_cast(
+        {:log, _src, _dest, <<0x23, 0x03, _ack_flag>>},
+        %{record_keepalives: false} = state
+      ) do
     {:noreply, state}
   end
 
-  def handle_cast({:log, binary, opts}, state) do
-    record = Record.new(binary, opts)
+  def handle_cast({:log, src, dest, binary}, state) do
+    record = Record.new(src, dest, binary)
+
+    remote_node = Record.remote_node(record)
+
+    state =
+      update_in(state.buffers[remote_node], fn
+        nil -> state.size |> CircularBuffer.new() |> CircularBuffer.insert(record)
+        buf -> CircularBuffer.insert(buf, record)
+      end)
 
-    {:noreply, %{state | buffer: CircularBuffer.insert(state.buffer, record)}}
+    {:noreply, state}
   end
 
   @impl GenServer
-  def handle_call(:clear, _from, %{buffer: %CircularBuffer{max_size: size}} = state) do
-    {:reply, :ok, %{state | buffer: CircularBuffer.new(size)}}
+  def handle_call(:clear, _from, state) do
+    {:reply, :ok, %{state | buffers: %{}}}
   end
 
-  def handle_call(:list, _from, %{buffer: buffer} = state) do
-    {:reply, CircularBuffer.to_list(buffer), state}
+  def handle_call(:list, _from, state) do
+    records =
+      for buffer <- Map.values(state.buffers),
+          record <- buffer,
+          do: record
```

Comment:
Uber nitpick just because I (personally) prefer using chains of `Enum` functions:
```suggestion
    records =
      state.buffers
      |> Map.values()
      |> Enum.concat()
```

Feel free to disregard!